src/Corax/Querying/Matches/SortingMatches/SortingMatch.BasicsComparers.cs

```diff
@@ -1,381 +1,216 @@
                 {
-                    return -(ca - cb);
+                    (usedBytes, usedChars) = ReadCharacter(_originalString, _stringBufferOffset, _curCharacters);
+                    curCharacterIsDigit = usedChars == 1 && char.IsDigit(_curCharacters[0]);
                 }
-
-                var cca = ToUpper(ca);
-                var ccb = ToUpper(cb);
-                if (cca != ccb)
+                else
                 {
-                    return cca - ccb;
+                    break;
                 }
-            }
-            else if (ca != cb)
-            {
-                return ca - cb;
-            }
-
-            ++ai;
-            ++bi;
+            } while (curCharacterIsDigit == _curSequenceIsNumber);
         }
 
-        NotAscii:
-        return CompareAlphanumericAscendingUtf8(a.Slice(ai), b.Slice(bi));
-    }
+        [ThreadStatic]
+        private static Decoder Decoder;
 
-    private static int CompareAlphanumericAscendingUtf8(ReadOnlySpan<byte> byteA, ReadOnlySpan<byte> byteB)
-    {
-        var auxiliarMemory = ArrayPool<char>.Shared.Rent(byteA.Length + byteB.Length);
-        var a = auxiliarMemory.AsSpan().Slice(0, byteA.Length);
-        var b = auxiliarMemory.AsSpan().Slice(byteA.Length, byteB.Length);
+        public AlphanumericStringComparisonState(ReadOnlySpan<byte> originalString, Span<char> curCharacters)
+        {
+            _curCharacters = curCharacters;
+            _originalString = originalString;
+            StringLength = (uint)originalString.Length;
+        }
 
-        var aLength = Encoding.UTF8.GetChars(byteA, a);
-        var bLength = Encoding.UTF8.GetChars(byteB, b);
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static (uint BytesUsed, uint CharUsed) ReadCharacter(ReadOnlySpan<byte> str, uint offset, Span<char> charactersBuffer)
+        {
+            var currentCharacter = Unsafe.Add(ref MemoryMarshal.GetReference(str), offset);
+            if ((uint)currentCharacter <= 0b0111_1111)
+            {
+                charactersBuffer[0] = (char)currentCharacter;
+                return (1, 1);
+            }
 
-        a.Slice(0, aLength);
-        b.Slice(0, bLength);
+            uint bytesUsed = ReadCharacterUtf8(str, offset, charactersBuffer, out uint charUsed);
+            return (bytesUsed, charUsed);
+        }
 
-        int result = 0;
-        int ai = 0, bi = 0;
-        while (true)
+        private static uint ReadCharacterUtf8(ReadOnlySpan<byte> str, uint offset, Span<char> charactersBuffer, out uint charUsed)
         {
-            char ca = ai < a.Length ? a[ai] : (char)0;
-            char cb = bi < b.Length ? b[bi] : (char)0;
+            var decoder = Decoder ??= Encoding.UTF8.GetDecoder();
 
-            /* skip over leading spaces or zeros */
-            while (IsSpace(ca))
-                ca = a[++ai];
+            //Numbers and ASCII are always 1 so we will pay the price only in case of UTF-8 characters.
```

Comment:
This can be done with SIMD instructions. I have a branch where I started building this routines.

Reply:
https://issues.hibernatingrhinos.com/issue/RavenDB-22675/Vectorize-methods-inside-Coraxs-alphanumerical-comparer.